rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,373 @@
+  Kibana has to provide an HTTP API to check users' permissions are sufficient to see a notification.
+  Unified Notification center performs a call to HTTP API to filter out notifications before rendering them in Cloud UI. 
+  This option adds a significant runtime performance penalty. Moreover, it makes the rendering process dependent on Kibana instance availability. If one of them is temporarily unavailable, the UNS can't decide if a notification must be shown or not.
+
+### Notification System
+Abstracts the way notifications are created, stored, configured, and retrieved.
+It allows Kibana to adopt different strategies for on-premise and Cloud versions:
+- *LocalRepository* - a repository for on-premise version, it stores data within the current deployment. 
+- *RemoteRepository* - a repository for the Cloud version of Kibana. It stores data within the Unified Notification System deployed in the Cloud.
+In this case, a user can see in Kibana UI the external notifications created outside of Kibana.
+
+*Repository* interface:
+- `create(NotificationEvent)`- creates a notification in the Notification system. 
+- `get({ search_after?: number; size?: number; type_id?: string[] })` - retrieves last notifications for the current user.
+Might be paginated using `search_after` filed.
+- `markAsRead({ notification_id: string[] })` - mark notifications with given ids as read.
+- `markAsUnread({ notification_id: string[] })` - mark notifications with given ids as unread.
+- `markAllAsRead()` - mark all the notifications as read.
+- `markAsPin({ notification_id: string[] })` - mark notifications with given ids as pinned.
+- `markAsUnpin({ notification_id: string[] })` - mark notifications with given ids as unpinned.
+- `configure()` - updates user-specific notification settings.
+
+#### LocalRepository
+![image](../images/0014/local_repository.png)
+
+##### Notification model
+Notification model is responsible for:
+- Applying policies to an incoming notification. NS must check whether a recipient unsubscribed from a source before processing to delivery.
+- Enhancing an incoming notification with necessary data: 
+    - `expired_at` if not specified. Kibana allows administrators to set how long notifications will be stored. Falls back to 30 - 90 days be default. 
+- Writing new incoming messages to the Notification storage.
+
+Processing a large number of incoming notifications might slow down Kibana and create a disproportionately high load on some Kibana instances.
+The use of a queue allows us to make the handling of incoming notifications asynchronously.
+Async mechanism lets Kibana to cope with spikes in the number of messages created. The queue implementation is based on the Elasticsearch server. 
+
+The logic is run as a background task. The usage of the single background task:
+- reduces the possibility of read- and write-conflicts
+- allows re-using task_manager built-in features: scheduling, retry, monitoring
```

Comment:
One potential snag: `task_manager` is currently in x-pack. I believe the recent licensing changes make this fine, but worth noting.

Reply:
Fair point. The answer depends on the Kibana structure after licensing changes https://github.com/elastic/kibana/pull/90297#discussion_r576193656 

Also, `task_manager` is a performance optimization. We can make it pluggable or implement it later (if necessary at all)

Reply:
I don't think that we should be using Task Manager here. Task Manager is capable of running a task every "x" seconds; however, this task is only run on one instance of Kibana. Since we are buffering these notifications in-memory prior to persisting them to disk, only one instance will get a chance to persist the notifications every "x" seconds.

Reply:
> Since we are buffering these notifications in-memory

Could you elaborate on why they are stored in memory? RFC suggests storing them in ES:
```
The use of a queue allows us to make the handling of incoming notifications asynchronous.
The async mechanism lets Kibana cope with spikes in the number of messages created. The queue implementation is based on the Elasticsearch server. 
```

Reply:
Perhaps I misunderstood the intent of the queue. I was under the impression that we wanted to use a queue to be able to "batch insert" new notifications so we didn't have a write operation per notification. Were you intending to use the ES based queue to read some persistent data and then create new data as a result on a periodic basis?

Reply:
> Were you intending to use the ES based queue to read some persistent data and then create new data as a result on a periodic basis?

Yes, new incoming messages are stored in ES in the form of a queue. A background task reads queued notifications, applies some logic to them, and stores notifications in the notifications storage.

Reply:
If we anticipate getting bursts of incoming messages, we should likely figure out a way to batch their inserts. Elasticsearch's [write queue is per operation](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-threadpool.html) at the moment until https://github.com/elastic/elasticsearch/issues/51336 is implemented.

Reply:
@kobelb Actually, I referred to a queue implementation based on Elasticsearch service. Something similar to the pattern that Reporting uses. After thinking a bit more on the topic, I don't think we should implement it shortly, as it complicates the system. We can come back to the matter when Kibana has more notifications that can lead to short spikes.
I'm going to remove the mention of the queue from the doc.